### PR TITLE
REF: de-duplicate groupby_helper code

### PR DESCRIPTION
**Pull request by jbrockmendel**
There's one other piece of de-duplication I think should be feasible but cython is still raising compilation errors for, so will do separately.

Orthogonal to #28931, but I expect it will cause merge conflicts.  28931 should be a higher priority.

### CODE REVIEW

File: pandas/_libs/groupby_helper.pxi.in [hunk 1]
```diff
@@ -19,6 +19,18 @@ ctypedef fused rank_t:
     object
 
 
+cdef inline bint _treat_as_na(rank_t val, bint is_datetimelike) nogil:
```
Comment (WillAyd) [inline pandas/_libs/groupby_helper.pxi.in]: Is this only applicable in groupby or should it go in util?

File: pandas/_libs/groupby_helper.pxi.in [hunk 2]
```diff
@@ -19,6 +19,18 @@ ctypedef fused rank_t:
...
+    if rank_t is object:
```
Comment (WillAyd) [inline pandas/_libs/groupby_helper.pxi.in]: Somewhat counter-intuitive that this works - out of curiosity what was the complaint?

File: pandas/_libs/groupby_helper.pxi.in [hunk 3]
```diff
@@ -59,24 +76,16 @@ def group_last(rank_t[:, :] out,
...
-                if rank_t is int64_t:
-                    # need a special notna check
-                    if val != NPY_NAT:
-                        nobs[lab, j] += 1
-                        resx[lab, j] = val
-                else:
-                    if val == val:
-                        nobs[lab, j] += 1
-                        resx[lab, j] = val
+                if val == val:
```
Comment (WillAyd) [inline pandas/_libs/groupby_helper.pxi.in]: Is this totally equivalent to what's in place? Looks like we are losing the `NPY_NAT` check?

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/groupby_helper.pxi.in]: Only here.  As I mention in comments below, I'm not wild about the fact that is_datetimelike is effectively hard-coded to True in several places.  Will try to see if that causes problems in follow-up(s)

File: pandas/_libs/groupby_helper.pxi.in (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/groupby_helper.pxi.in]: AFAICT the `val != val` check requires calling `val.__ne__` which requires the gil

File: pandas/_libs/groupby_helper.pxi.in (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/groupby_helper.pxi.in]: This is in the `rank_t is object` case where we cant use _treat_as_na.  (comment just below here)

File: pandas/_libs/groupby_helper.pxi.in (hunk 2 above)
Comment (jreback) [reply pandas/_libs/groupby_helper.pxi.in]: shouldn’t need the gil if these are c level object (eg ints or floats)

Comment (jbrockmendel) [reply pandas/_libs/groupby_helper.pxi.in]: Right, we just need to exclude the object case from getting to the `val == val` step, thats all this check is doing

File: pandas/_libs/groupby_helper.pxi.in [hunk 4]
```diff
@@ -19,6 +19,18 @@ ctypedef fused rank_t:
     object
 
 
+cdef inline bint _treat_as_na(rank_t val, bint is_datetimelike) nogil:
+    if rank_t is object:
+        # Should never be used, but we need to avoid the `val != val` below
+        #  or else cython will raise about gil acquisition.
```
Comment (jreback) [inline pandas/_libs/groupby_helper.pxi.in]: this is pretty odd, why dont you return an enum here (true, false, raise) and handle in code appropriately.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/groupby_helper.pxi.in (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/groupby_helper.pxi.in]: the vicissitudes of cython fused types

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm any perf implications?

**Comment by jbrockmendel** [comment]
Rebased + green.

**Comment by WillAyd** [comment]
Thanks @jbrockmendel

**Pull request merged**